features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,318 @@
+# Objective
+
+- Provide cached result to Chat Container API calls
+- Compatible to old API interface
+
+# Overview
+
+Container contains a local store for caching resources. The local store caches resources by resource id.
+
+### Store Update
+
+When SDK recieves result from server, for each record, the newly fetched record would always replace the old one with the same record id if there is one.
+
+### API
+
+For fetch API of each resource, there would be a new callback that should return result from cache immediately.
+
+The interface for each language may be different.
+
+```
+getResource(params, callback, cachedCallback?)
+```
+
+Resource includes conversation and message.
+
+The cachedCallback should be optional, thus compatible to the old API interace.
+
+After client app receives result from both callback, client app need to resolve a single result for display. Skygear Chat Container would provide some utility functions to help client app resolve two callback result.
+
+### Expected UI Display / Common Use Case
+
+Client App is expected to keep a single list of messages for display in table. And they are required to keep track to the newest and oldest fetched messages to determine when to fetch result from server.
+
+Client App should replace old message with new message if the same message id is found.
+
+Deleted messages are also returned in fetch API, with empty content and marked as deleted. This is to keep the whole message list complete. Client App is expected to skip these message in UI display.
+
+We may provide a high level API **ManagedMessageList** to handle this for the developer.
+
+# Sample Code for Message List, with ManagedMessageList
+
+*All below UI update, I use `this.tableView.reloadTable()`. It is up to client app to decide how to update the UI*
+
+#### Enter conversation
+
+```
+this.stickToBottom = true | false
+this.messageList = new ManagedMessageList(container: skygear.defaultContainer(),
+                                          conversation: this.conversation,
+                                          delegate: this)
+
+this.messageList.fetchLatestMessages()
```

Comment:
Why users need to call `fetchLatestMessage()`? There are no return value and parameter which makes me wonder why it was not in the constructor or...?

Reply:
I want to leave some flexibility to `ManagedMessageList`.

I expect the managed list contains the logic of how to resolve cached result and fetched result.

The scrollViewDelegate implementation is a bonus to the list to also contains the logic of when to fetch what.

So the reason why `fetchLatestMessage` does not return value is that the value are handled by the managed list. 

And then the list itself without attaching to scroll view delegate does not call any api, so user need to call it themself.

I expect a more complicated version of the `ManagedMessageList` in the future can also accept a list of messages in constructor, and developers can choose to fetch messages before/after those messages before even entering the message list.

This gives some flexibility to the developer to use the managed list to display messages within any time frame that they want.